src/api/db/data/20200702145811_adapt_review_and_duplicated_notifications.rb

```diff
@@ -0,0 +1,64 @@
+class AdaptReviewAndDuplicatedNotifications < ActiveRecord::Migration[6.0]
+  def up
+    transform_review_notifications
+    remove_outdated_notifications
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
+  end
+
+  private
+
+  # associate the review notifications to the corresponding bs_request, since
+  # they won't be treated separately anymore
+  def transform_review_notifications
+    Notification.where("notifiable_type = 'Review'").find_each do |notification|
+      notification.update(notifiable_id: notification.notifiable.bs_request.id, notifiable_type: 'BsRequest')
+    end
+  end
+
+  # only keep the latest comment and request notifications related to the same notifiable and subscriber
+  def remove_outdated_notifications
+    Notification.find_each do |notification|
+      next if delete_notification_without_notifiable(notification)
+
+      notifications = outdated_notifications(notification)
+
+      next if notifications.count <= 1
+
+      # check for the highest id, because the 'created_at' date can be the same for multiple notifications,
+      # since a delayed_job is creating them in batches
+      newest_notification = notifications.first
+      notifications.where.not(id: newest_notification.id).destroy_all
```

Comment:
I think you can use `notifications.reject { |n| n.id == newest_notification.id }.destroy_all` and avoid querying the DB again with the `where`.

Reply:
You can try also this
```ruby
newest_notification = notifications.shift
notifications.destroy_all
```

Reply:
Hmmm, `shift` doesnt work on `ActiveRecord_Relation` objects, the `reject` returns an array and you can't do the `destroy_all` on it. I will check if what we can do here

Reply:
I know just iterate over the already fetched records, and call `.destroy`. This avoids querying the two times. `destroy_all` does the same, it calls `destroy` on every record. Couldn't find a more elegant solution that doesn't cause more complexity.

Reply:
okay, we changed it so we use `.offset(1)` on the queries in `fetch_comment_notifications` and fetch_request_notifications`. This way we do not include the newest notification into the collection, and we can simply call `destroy_all` on them

Reply:
https://guides.rubyonrails.org/v6.0/active_record_querying.html#limit-and-offset

Reply:
Awesome!